core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -188,23 +189,6 @@ public static <T extends PartitionScanTask> List<ScanTaskGroup<T>> planTaskGroup
     return taskGroups;
   }
 
-  private static StructLike projectGroupingKey(
-      StructProjection groupingKeyProjection,
-      Types.StructType groupingKeyType,
-      StructLike partition) {
-
-    PartitionData groupingKey = new PartitionData(groupingKeyType);
```

Comment:
This line was causing the issue.

Reply:
Just for my understanding, the repeated calling of this constructor was expensive because the repeated data array initialization (and possibly the determining of the partitionType via `AvroSchemaUtil.convert`)? And now we avoid repeatedly performing those computations by having a `PartitionData` be initialized with a "template" just once for all tasks and then replace the actual data in `PartitionData`?

Reply:
Yep, correct, @amogh-jahagirdar.